### ISSUE TITLE
Remember debug settings

### ISSUE THREAD

**Issue by Adamduehansen** [excaliburjs/excalibur-extension#8]
Hi.

I was looking at ways to contibute on the browser extension since it such a nice tool and I have some experience with working on extensions. On the wish list is an item called "Remember debug settings across refreshes/closing". I thought this might be an easy way to get into working on the extension and a quality of life improvement for users.

I was thinking that a simple way to implement this would be to use the storage API for extensions. Have there been any other thoughts on this?

**Comment by eonarheim** [comment]
Awesome, this sounds like a great idea. I haven't attempted to save
settings yet so poking at the storage api seems like a good start.

**Comment by Adamduehansen** [comment]
All right. I will take a look at an implementation.

**Comment by Adamduehansen** [comment]
@eonarheim Is the default settings object declared like this
```ts
export const DefaultSettings: Settings = {
    showNames: true,
    showIds: false,
    showPos: false,
    showPosLabel: false,
    posColor: {r: 0, g: 0, b: 0, a: 1},
    showGraphicsBounds: false,
    graphicsBoundsColor: {r: 0, g: 0, b: 0, a: 1},
    showColliderBounds: false,
    colliderBoundsColor: {r: 0, g: 0, b: 0, a: 1},
    showGeometryBounds: false,
    geometryBoundsColor: {r: 0, g: 0, b: 0, a: 1}
}
```
because it is the default debug settings for Excalibur? And changing a boolean on the `DefaultSettings` object will have no effect on what values are actually used?

As I understand it the actual settings is read from the `window.___EXCALIBUR_DEVTOOL.game.debug` property during the "heartbeat" event? 🙂

**Comment by eonarheim** [comment]
> @eonarheim Is the default settings object declared like this
> 
> ```ts
> export const DefaultSettings: Settings = {
>     showNames: true,
>     showIds: false,
>     showPos: false,
>     showPosLabel: false,
>     posColor: {r: 0, g: 0, b: 0, a: 1},
>     showGraphicsBounds: false,
>     graphicsBoundsColor: {r: 0, g: 0, b: 0, a: 1},
>     showColliderBounds: false,
>     colliderBoundsColor: {r: 0, g: 0, b: 0, a: 1},
>     showGeometryBounds: false,
>     geometryBoundsColor: {r: 0, g: 0, b: 0, a: 1}
> }
> ```
> 
> because it is the default debug settings for Excalibur? And changing a boolean on the `DefaultSettings` object will have no effect on what values are actually used?
> 
> As I understand it the actual settings is read from the `window.___EXCALIBUR_DEVTOOL.game.debug` property during the "heartbeat" event? 🙂

Good question, I needed to do a little research! You are correct I think the initial setting is actually hydrated from the running game https://github.com/excaliburjs/excalibur-extension/blob/main/src/background.js#L63

I think the only purpose of the `DefaultSettings` is to give the component defaults in cause of error so it renderers

**Comment by Adamduehansen** [comment]
All right, thank you for the clarification. It makes sense.

I was thinking that the `firstUpdated` method inside "app-main.ts" could be used to load the settings through the storage API. Then an event could be dispatched to the background script through `this.backgroundConnection`, which inserts the debug settings into the Excalibur instance. The updated debug settings are then hydrated to the extension on next the "heartbeat" event.

I'll try to write a PR on an implementation like this.

**Comment by eonarheim** [comment]
@Adamduehansen That seems very reasonable to me given the current state of the plugin, thanks for digging into this!

**Comment by Adamduehansen** [comment]
Hi @eonarheim . I have been looking into this issue for the last few weeks. Unfortunately, it wasn't quite as easy as I had hoped. I might have underestimated how the application worked. To get this feature to work there seems to be two issues:
* The devtool panel loses its connection to the background service on refreshes, which means that the dev tools panel cannot send updates to the background script.
* I think that a rewrite of the background script is required since the “main” world, where the heartbeat is fired, does not have access to the extension storage API. The background script would instead be the one firing a heatbeat to the “main” world with the settings.

With the knowledge of how this extension works, I have been able to create a new small proof of concept extension that solves the above points and can store settings between refreshes. However I’m unsure how to proceed from here. I would like to help with the further development of this extension overall but I feel like some refactoring and cleanup is needed before this issue can be solved.

I can send you a repo with my proof of concept so that you can try it yourself?

**Comment by eonarheim** [comment]
>  Unfortunately, it wasn't quite as easy as I had hoped.

@Adamduehansen thanks again for digging into this! Browser extensions are a new area for me, so any findings are incredibly valuable.

> I can send you a repo with my proof of concept so that you can try it yourself?

Totally! I'd love to see it, you can DM it to me or shoot me a public repo.

> I think that a rewrite of the background script is required since the “main” world, where the heartbeat is fired, does not have access to the extension storage API. The background script would instead be the one firing a heatbeat to the “main” world with the settings.

Interesting, let's noodle. Admittedly I'm an extension developer noob, so I'm willing to consider changes. Extensions seem to have frustrating rules and confusing documenttion that make dev very complicated unfortunately.

**Comment by eonarheim** [comment]
You were looking into this api? https://developer.chrome.com/docs/extensions/reference/api/storage

**Comment by Adamduehansen** [comment]
> Extensions seem to have frustrating rules and confusing documenttion that make dev very complicated unfortunately.

Exactly. I’ve been working on a few browser extensions at work, so by now I kinda know how to navigate it. Usually only for Chrome though.

> Totally! I'd love to see it, you can DM it to me or shoot me a public repo.

I might have something better. I had some free time during this weekend so I went ahead and implemented the ideas from the POC onto this extension. So far it seems to work in that debug settings are re-applied after page refreshes and even when the dev panel is closed and re-opened. My plan is to develop a small game during this week with this branch of the extension to check if nothing is broken.
You can pull my fork and test it as well: https://github.com/Adamduehansen/excalibur-extension/tree/feat/extension-settings

> You were looking into this api? https://developer.chrome.com/docs/extensions/reference/api/storage

Yeah but I did not use it for this implementation. The issue is that the API is not accessible where the loop is running. The way I got around this is by moving it to the background script.

The settings will be stored until the browser is closed, but I think that this could be fixed in the future if needed 🙂

**Comment by Adamduehansen** [comment]
Also, I've noticed that there is something going on build process, since not all files are build during build scripts. I have no experience with Parcel as a bundler so I don't know what's going on.
